### PR TITLE
Random dates + test

### PR DESCRIPTION
**Pull request by lizlouise1335**
Random dates generator + generator tests

### CODE REVIEW

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import numpy as np
+import pandas as pd
+from numpy.random import Generator
+from typing import Optional
+
+
+def generate_datetime(
```
Comment (ksneab7) [inline synthetic_data/dataset_generators/datetime_generator.py:7]: I would recommend putting the default values for the start and end date variables in function's doc string, just because if we ever want to autogenerate documentation it would be good to have in there already

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    date_list = []
+    if not date_format_list:
+        date_format_list = ["%B %d %Y %H:%M:%S"]
+
+    for _ in range(num_rows):
+        date_format = rng.choice(date_format_list)
+        datetime = generate_datetime(
+            rng, date_format=date_format, start_date=start_date, end_date=end_date
+        )
+        date_list.append(datetime)
```
Comment (JGSweets) [inline synthetic_data/dataset_generators/datetime_generator.py:77]: I wonder if we can pre-allocate a list to reduce the expense of the append.
something like:

```python
date_list = ["%B %d %Y %H:%M:%S"] * num_rows

and then below, use the index to insert in the correct place.
```

There might even be more interesting optimizations that could make this overall faster instead of looping each row.
But we can optimize that in a future PR if desired.

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    return np.array(date_list)
+
+
+
+
+
```
Comment (taylorfturner) [inline synthetic_data/dataset_generators/datetime_generator.py:84]: no need for the 4 extra lines EOF (end of file)

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import pandas as pd
+from numpy.random import Generator
+from typing import Optional
+
+
+def generate_datetime(
+    rng: Generator,
+    date_format: str,
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
```
Comment (JGSweets) [inline synthetic_data/dataset_generators/datetime_generator.py:11]: I don't believe these are `str` given usage below:
```python
start_date = pd.Timestamp(1920, 1, 1)
```

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    :return: generated datetime
+    :rtype: str
+    """
+    if not start_date:
+        # 100 years in past
+        start_date = pd.Timestamp(1920, 1, 1)
+    if not end_date:
+        # protection of 30 years in future
+        end_date = pd.Timestamp(2049, 12, 31)
```
[the remark this answers is not in the document]
Comment (JGSweets) [reply synthetic_data/dataset_generators/datetime_generator.py:36]: What's the mutability of that though. Might be better to just notate in the docstring

File: tests/test_datetime_generator.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            format_check = False
+        self.assertTrue(format_check)
+
+    def test_generate_datetime_range(self):
+        date_str = dtg.generate_datetime(self.rng, 
+                                        self.date_format_list[0],
+                                        self.start_date,
+                                        self.end_date)
+        date_obj = pd.to_datetime(date_str, format = self.date_format_list[0])
+        self.assertTrue(self.start_date <= date_obj <= self.end_date)
```
Comment (micdavis) [inline tests/test_datetime_generator.py:43]: careful, I think this compound inequality works in python, but in a lot of languages this will not equate as expected due to order of execution for operators.

File: synthetic_data/dataset_generators/datetime_generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JGSweets) [reply synthetic_data/dataset_generators/datetime_generator.py:36]: Thinking more: `pd.Timestamp(1920, 1, 1)` might be fine bc it is defined every time. Need to think more about it.

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    :param start_date: the latest date that datetimes can be generated at,
+        defaults to None
+    :type start_date: pd.Timestamp, None, optional
+
+    :return: array of generated datetimes
+    :rtype: numpy array
+    """
+    date_list = []
+    if not date_format_list:
+        date_format_list = ["%B %d %Y %H:%M:%S"]
```
[the remark this answers is not in the document]
Comment (micdavis) [reply synthetic_data/dataset_generators/datetime_generator.py:70]: you can ignore this because I didn't realize that mutable input args should default to None.

File: tests/test_datetime_generator.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                                        self.end_date)
+        format_check = True
+        try:
+            pd.to_datetime(date_str, format=self.date_format_list[0])
+        except ValueError:
+            format_check = False
+        self.assertTrue(format_check)
+
+    def test_generate_datetime_range(self):
+        date_str = dtg.generate_datetime(self.rng,
```
Comment (JGSweets) [inline tests/test_datetime_generator.py:38]: I'm curious how numpy.random asserts their tests. Might be worth to get considerations of best practices when asserting for random generators.

Comment (JGSweets) [reply tests/test_datetime_generator.py:38]: https://github.com/numpy/numpy/blob/main/numpy/random/tests

File: synthetic_data/dataset_generators/datetime_generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lizlouise1335) [reply synthetic_data/dataset_generators/datetime_generator.py:36]: Timestamps are immutable, should I change anything or leave as is for now?

File: tests/test_datetime_generator.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        date_str = dtg.generate_datetime(self.rng, 
+                                        self.date_format_list[0],
+                                        self.start_date,
+                                        self.end_date)
+        format_check = True
+        try:
+            pd.to_datetime(date_str, format=self.date_format_list[0])
+        except ValueError:
+            format_check = False
+        self.assertTrue(format_check)
```
[the remark this answers is not in the document]
Comment (lizlouise1335) [reply tests/test_datetime_generator.py]: that worked

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    :param end_date: the latest date that datetimes can be generated at,
+        defaults to pd.Timestamp(2049, 12, 31)
+    :type end_date: pd.Timestamp, None, optional
+
+    :return: generated datetime
+    :rtype: str
+    """
+    if not start_date:
+        # 100 years in past
+        start_date = pd.Timestamp(1920, 1, 1)
```
Comment (JGSweets) [inline synthetic_data/dataset_generators/datetime_generator.py:33]: if we keep these, we should validate these edge conditions in tests

File: synthetic_data/dataset_generators/datetime_generator.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    :param end_date: the latest date that datetimes can be generated at,
+        defaults to pd.Timestamp(2049, 12, 31)
+    :type end_date: pd.Timestamp, None, optional
+
+    :return: array of generated datetimes
+    :rtype: numpy array
+    """
+    date_list = [""] * num_rows
+    if not date_format_list:
+        date_format_list = ["%B %d %Y %H:%M:%S"]
```
Comment (JGSweets) [inline synthetic_data/dataset_generators/datetime_generator.py:70]: We should validate this edge condition in tests

File: tests/test_datetime_generator.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+                                        self.date_format_list[0],
+                                        self.start_date,
+                                        self.end_date)
+        date_obj = pd.to_datetime(date_str, format = self.date_format_list[0])
+        self.assertTrue(self.start_date <= date_obj)
+        self.assertTrue(date_obj <= self.end_date)
+
+    @mock.patch("synthetic_data.dataset_generators.datetime_generator.generate_datetime")
+    def test_random_datetimes_return_type_and_size(self, mock_generate_datetime):
+        mock_generate_datetime.return_value = "Mocked"
```
Comment (JGSweets) [inline tests/test_datetime_generator.py:46]: if we want to mock out `generate_datetime` bc of the randomness, I suggest asserting what it is called with. My main concern rn is we aren't really validating it will select from a list. 

A different way to make it deterministic, set `self.rng = mock.Mock(spec=....)` and then set the return value for the usages. We assume the np random functions work as expected.

e.g.
```python
mocked_rng = ...
mocked_rng.random.return_value = 0.5

# or use a side_effect + generator so it can change values each use
mocked_rng.random.side_effect = (x for x in [0.25, 0.5, 0.75])  # gen with 3 uses
```

Now we know what `t` will always be, we know the start \ end times, and we can do the same for formats with `choice` so we can ensure given the right inputs we know the expected output. Definitely a bit more work and maybe overkill.

File: tests/test_datetime_generator.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
+
+
```
Comment (JGSweets) [inline tests/test_datetime_generator.py:2]: let's remove these spaces

**Review by JGSweets — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]
